lib/Doctrine/Annotations/Type/CompositeType.php

```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Annotations\Type;
+
+/**
+ * @internal
+ */
+interface CompositeType extends Type
```

Comment:
Should be replaced with abstract class so we stop duplicating logic in Union and Intersection types

Reply:
It's not duplicated logic, intersecions and unions are opposite patterns. Even then I'd prefer trait.

Reply:
So what is this if not duplicated logic

<img width="1337" alt="Screenshot 2019-03-31 at 20 27 52" src="https://user-images.githubusercontent.com/496233/55293153-811f9600-53f3-11e9-95ec-30ace13bd593.png">

Reply:
But it's still not a good argument for using inheritance

Reply:
Describe maybe. validate() is not duplicated logic.

Reply:
validate() is duplicated logic with inverted result. Equivalent to `return ! parent::validate($value)`

I don't care about inheritance but duplicated logic. Fix in one class will require touching both files. Use your favorite composition if you think it's more appropriate, whatever. You can also use traits, I don't care.

Reply:
Extracted describe() into static utility function but keeping validate as is.
`return ! parent::validate($value)` is not valid approach and using such would make the code less understendable. Feel free to come up with follow-up PR if you find nice solution. 👍